File: examples/nodejs/geocode.js
/**
 * CSV2GEO API - Node.js Geocoding Example
 */

const API_KEY = 'YOUR_API_KEY';
const BASE_URL = 'https://api.csv2geo.com/v1';

/**
 * Forward geocode an address to coordinates
 * @param {string} address - The address to geocode
 * @returns {Promise<object>} Geocoding result
 */
async function geocode(address) {
  const params = new URLSearchParams({
    q: address,
    api_key: API_KEY
  });

  const response = await fetch(`${BASE_URL}/geocode?${params}`);

  if (!response.ok) {
    throw new Error(`Geocoding failed: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Reverse geocode coordinates to an address
 * @param {number} lat - Latitude
 * @param {number} lng - Longitude
 * @returns {Promise<object>} Address result
 */
async function reverseGeocode(lat, lng) {
  const params = new URLSearchParams({
    lat: lat.toString(),
    lng: lng.toString(),
    api_key: API_KEY
  });

  const response = await fetch(`${BASE_URL}/reverse?${params}`);

  if (!response.ok) {
    throw new Error(`Reverse geocoding failed: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Batch geocode multiple addresses
 * @param {string[]} addresses - Array of addresses (max 10,000)
 * @returns {Promise<object>} Batch results
 */
async function batchGeocode(addresses) {
  const response = await fetch(`${BASE_URL}/geocode`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${API_KEY}`
    },
    body: JSON.stringify({ addresses })
  });

  if (!response.ok) {
    throw new Error(`Batch geocoding failed: ${response.statusText}`);
  }

  return response.json();
}

// Example usage
async function main() {
  try {
    // Forward geocoding
    console.log('Forward Geocoding:');
    const result = await geocode('1600 Pennsylvania Ave, Washington DC');
    console.log(`  Address: ${result.results[0].formatted_address}`);
    console.log(`  Lat: ${result.results[0].location.lat}`);
    console.log(`  Lng: ${result.results[0].location.lng}`);

    // Reverse geocoding
    console.log('\nReverse Geocoding:');
    const reverseResult = await reverseGeocode(38.8977, -77.0365);
    console.log(`  Address: ${reverseResult.results[0].formatted_address}`);

    // Batch geocoding
    const addresses = [
      '1600 Pennsylvania Ave, Washington DC',
      '350 Fifth Avenue, New York, NY',
      '1 Infinite Loop, Cupertino, CA'
    ];
    const batchResult = await batchGeocode(addresses);
    console.log(`\nBatch Geocoding: ${batchResult.results.length} addresses processed`);

  } catch (error) {
    console.error('Error:', error.message);
  }
}

main();

File: examples/python/geocode.py
"""
CSV2GEO API - Python Geocoding Example
"""
import requests

API_KEY = "YOUR_API_KEY"
BASE_URL = "https://api.csv2geo.com/v1"


def geocode(address: str) -> dict:
    """
    Forward geocode an address to coordinates.

    Args:
        address: The address to geocode

    Returns:
        dict: Geocoding result with lat/lng
    """
    response = requests.get(
        f"{BASE_URL}/geocode",
        params={
            "q": address,
            "api_key": API_KEY
        }
    )
    response.raise_for_status()
    return response.json()


def reverse_geocode(lat: float, lng: float) -> dict:
    """
    Reverse geocode coordinates to an address.

    Args:
        lat: Latitude
        lng: Longitude

    Returns:
        dict: Address result
    """
    response = requests.get(
        f"{BASE_URL}/reverse",
        params={
            "lat": lat,
            "lng": lng,
            "api_key": API_KEY
        }
    )
    response.raise_for_status()
    return response.json()


def batch_geocode(addresses: list[str]) -> dict:
    """
    Batch geocode multiple addresses.

    Args:
        addresses: List of addresses (max 10,000)

    Returns:
        dict: Batch geocoding results
    """
    response = requests.post(
        f"{BASE_URL}/geocode",
        headers={"Authorization": f"Bearer {API_KEY}"},
        json={"addresses": addresses}
    )
    response.raise_for_status()
    return response.json()


if __name__ == "__main__":
    # Example: Forward geocoding
    result = geocode("1600 Pennsylvania Ave, Washington DC")
    print("Forward Geocoding:")
    print(f"  Address: {result['results'][0]['formatted_address']}")
    print(f"  Lat: {result['results'][0]['location']['lat']}")
    print(f"  Lng: {result['results'][0]['location']['lng']}")

    # Example: Reverse geocoding
    result = reverse_geocode(38.8977, -77.0365)
    print("\nReverse Geocoding:")
    print(f"  Address: {result['results'][0]['formatted_address']}")

    # Example: Batch geocoding
    addresses = [
        "1600 Pennsylvania Ave, Washington DC",
        "350 Fifth Avenue, New York, NY",
        "1 Infinite Loop, Cupertino, CA"
    ]
    result = batch_geocode(addresses)
    print(f"\nBatch Geocoding: {len(result['results'])} addresses processed")
